library/java/src/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java

```diff
@@ -0,0 +1,99 @@
+package io.envoyproxy.envoymobile.engine;
+
+import android.annotation.TargetApi;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.net.ConnectivityManager;
+import android.net.ConnectivityManager.NetworkCallback;
+import android.net.Network;
+import android.net.NetworkCapabilities;
+import android.net.NetworkInfo;
+import android.net.NetworkRequest;
+import android.os.Build;
+
+/**
+ * This class makes use of some deprecated APIs, but it's only current purpose is to attempt to
+ * distill some notion of a preferred network from the OS, upon which we can assume new sockets will
+ * be opened.
+ */
+@TargetApi(Build.VERSION_CODES.LOLLIPOP)
+public class AndroidNetworkMonitor extends BroadcastReceiver {
+  private static final int ENVOY_NET_GENERIC = 0;
+  private static final int ENVOY_NET_WWAN = 1;
+  private static final int ENVOY_NET_WLAN = 2;
+
+  private static volatile AndroidNetworkMonitor instance = null;
+
+  private ConnectivityManager connectivityManager;
+  private NetworkCallback networkCallback;
+  private NetworkRequest networkRequest;
+
+  public static void load(Context context) {
+    if (instance != null) {
+      return;
+    }
+
+    synchronized (instance) {
+      if (instance != null) {
+        return;
+      }
+      instance = new AndroidNetworkMonitor(context);
+    }
+  }
+
+  private AndroidNetworkMonitor(Context context) {
+    connectivityManager =
+        (ConnectivityManager)context.getSystemService(Context.CONNECTIVITY_SERVICE);
+    networkRequest = new NetworkRequest.Builder()
+                         .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
+                         .build();
+
+    networkCallback = new NetworkCallback() {
+      @Override
+      public void onAvailable(Network network) {
+        handleNetworkChange();
+      }
+      @Override
+      public void onCapabilitiesChanged(Network network, NetworkCapabilities networkCapabilities) {
+        handleNetworkChange();
+      }
+      @Override
+      public void onLosing(Network network, int maxMsToLive) {
+        handleNetworkChange();
+      }
+      @Override
+      public void onLost(final Network network) {
+        handleNetworkChange();
+      }
+    };
+
+    context.registerReceiver(this, new IntentFilter() {
+      { addAction(ConnectivityManager.CONNECTIVITY_ACTION); }
+    });
+  }
+
+  @Override
+  public void onReceive(Context context, Intent intent) {
+    handleNetworkChange();
+  }
+
+  private void handleNetworkChange() {
+    NetworkInfo networkInfo = connectivityManager.getActiveNetworkInfo();
+    if (networkInfo == null) {
+      AndroidJniLibrary.setPreferredNetwork(ENVOY_NET_GENERIC);
```

Comment:
It feels a bit too loose for the `setPreferredNetwork()` be a static method since it depends on the `AndroidJniLibrary` to be loaded first.

What do you think about making it an instance method of `AndroidJniLibrary` and we can have this `AndroidNetworkMonitor` depend on `AndroidJniLibrary` to call this?

Reply:
I think that's a valid concern, but it sort of applies equally to _any_ of the current JNI library methods, doesn't it?

I can get behind making both JniLibrary classes instance-based (this could make testing easier too), but would you be alright with having that be a separate issue/PR?

Reply:
Yep! Let's do it in a separate PR